net-print/cnijfilter: new package version 3.80

mfld-fr opened the pull request:
Add support of the Canon IP 7200 series inkjet printers to the CUPS subsystem.

Contains actually two packages : the common part to all printers (cnijfilter-common), and the printer-specific part (cnijfilter-ip7200). This layout was requested by Gentoo developers, to allow upgrading the common part while keeping the printer-specific part that does not provide downward compatibility between versions.

The sources are still provided by the manufacturer, and that support is frequently requested in Bugzilla
(see linked bugs).

Gentoo-Bug: https://bugs.gentoo.org/130645

Built on AMD64 architecture, and tested with IP 7250 printer model on Wifi. There are still several but without impact compilation warnings that would be fixed in a next release.

This package was sourced from the Sabayon overlay, that still proposes it for mainstream (see https://github.com/Sabayon/for-gentoo/pull/69), then tuned to pass the repoman check.

Package-Manager: Portage-2.3.6, Repoman-2.3.1

mfld-fr commented:
Rebased the cnijfilter-1 branch as requested.

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+RDEPEND="
+	net-print/cnijfilter[servicetools?]
+	>=media-libs/libpng-1.5
```
libpng is slotted, so you probably want `:0=`.

mfld-fr replied:
Fixed

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+RDEPEND="
+	net-print/cnijfilter[servicetools?]
+	>=media-libs/libpng-1.5
+	>=media-libs/tiff-3.4
```
Likewise.

mfld-fr replied:
Fixed

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+DESCRIPTION="Canon InkJet Printer Driver for Linux (Pixus/Pixma-Series)."
+HOMEPAGE="http://support-sg.canon-asia.com/contents/SG/EN/0100469302.html"
+SRC_URI="http://gdlp01.c-wss.com/gds/3/0100004693/01/${MY_PN}-source-${PV}-1.tar.gz"
```
You're using the filename twice (here and in S), so put it into `MY_P` variable instead. You can also kill `MY_PN` since it is of no value as separate variable.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+pkg_setup() {
+	[[ -z ${LINGUAS} ]] && LINGUAS="en"
+
+	DIRS_PRINTER="cnijfilter"
```
Prefer bash arrays over appending to string lists.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+	# missing macros directory make aclocal fail
+	mkdir maintenance/m4 || die
+
+	epatch \
```
epatch is deprecated, use apply.

mfld-fr replied:
Fixed.

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+	mkdir maintenance/m4 || die
+
+	epatch \
+		"${FILESDIR}/${MY_PN}"-3.70-png.patch \
```
You have full control over patch names, so don't make things more complex by using MY_PN here.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+S="${WORKDIR}/${MY_PN}-source-${PV}-1"
+
+_dir_build() {
```
This looks like a lot of indirection for little value. Kill it.

mfld-fr replied:
Fixed

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+		pushd ${d} >/dev/null
+		# progpath must be set otherwise we go for /usr/local/bin
+		if [[ ${command} == econf* ]]; then
+			${command} --enable-progpath="${EPREFIX}/usr/bin"
```
By the way, this hack is horrible. It really does not belong in this function.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+	for (( i=0; i<${#PRINTER_USE[@]}; i++ )); do
+		local name="${PRINTER_USE[$i]}"
+		local pid="${PRINTER_ID[$i]}"
```
This could be a bit cleaner if you used a single list with values separated by some char instead of two, e.g. something like:

    PRINTERS=( model:id model:id model:id )
    IUSE="${PRINTERS[@]%:*} ..."

    ...
    for p in "${PRINTERS[@]}"; do
      local name=${p%:*}
      local pid=${p#*:}

Most notably, this makes it easier to figure out which is which, without having to match indexes.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+		local pid="${PRINTER_ID[$i]}"
+		if use ${name}; then
+			mkdir -p ${name} || die
+			ln -s "${S}"/${pid} ${name}/ || die
```
Why do you need to use absolute path here?

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+	done
+}
+
+_printer_dir_build() {
```
Like the other helper.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+			mkdir -p ${name} || die
+			ln -s "${S}"/${pid} ${name}/ || die
+			for d in ${DIRS_PRINTER}; do
+				cp -a ${d} ${name} || die
```
This whole logic seems like reinventing the wheel of multibuild.eclass. If you need to build the sources for every model separately, reuse the logic already in the eclass instead of reinventing your own, poorly.

mfld-fr replied:
Could you please elaborate ? I know that `multilib` is used to build variants from the same source, but I don't know how to use it to iterate over different sources (= sub-directories here).

mgorny replied:
Look at multibuild.eclass. It's used to build N different copies (variants) of the same package.

mfld-fr replied:
OK, but in that case, we don't want to build variants of the same package. We have several subdirectories in the source package, each of them is a component having its own configure and makefile, so we have to iterate over that subdirectories (and select some of them depending on USE flags). You wrote that the helper function _dir_build() is ugly, but what do you suggest as a cleaner way to iterate over that subdirectories ?

mgorny replied:
But in this step, you are creating a copy of every input directory for every printer enabled, am I correct? This is what multibuild is for.

For using those directories afterwards, just inline a trivial for loop, without any fancy prints.

mgorny replied:
By the way, try to build out-of-source if possible; i.e. instead of copying those directories use build dirs with `ECONF_SOURCE=${S}/${dir} econf`.

mfld-fr replied:
OK, got it ! Yes, this source copy inherited from Sabayon is ugly and I will try to improve that part when reworking the `cnijfilter-drivers` package for the second commit. My question is more on the current `cnijfilter` package and its _dir_build() helper. What do you suggest to make it cleaner than in my latest intermediate commit ?

mfld-fr replied:
Closed as TODO in `cnijfilter-drivers` ebuild.

mgorny commented on net-print/cnijfilter-drivers/files/cnijfilter-3.70-libexec-cups.patch:
```diff
@@ -0,0 +1,9 @@
+diff -urN cnijfilter-source-3.70-1.old/pstocanonij/filter/Makefile.am cnijfilter-source-3.70-1/pstocanonij/filter/Makefile.am
```
Combine the patches that do the same thing. Also, would be nice to have some explanation what a patch does on the top and why.

mfld-fr replied:
Well... we could combine these patches, for sure, but I am currently in the phase of recovering as many proposed patches as possible from many guys in many locations, and I would like to keep the traceability by keeping the patch names and contents in that phase. Otherwise it would become practically impossible to consolidate that long history.

mfld-fr replied:
Closed

mgorny commented on net-print/cnijfilter-drivers/metadata.xml:
```diff
@@ -0,0 +1,22 @@
+    <email>mfld.fr@gmail.com</email>
+    <name>mfld-fr</name>
+  </maintainer>
+  <longdescription>
```
Add the proxy-maint team as second maintainer so there will be someone to commit for you.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+inherit eutils autotools flag-o-matic multilib
+
+DESCRIPTION="Canon InkJet Printer Driver for Linux (Pixus/Pixma-Series)."
```
This is not a proper sentence, so no `.` at the end.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+
+MY_PN="${PN/-drivers/}"
+
+inherit eutils autotools flag-o-matic multilib
```
Unless I'm missing something, you're not using `eutils flag-o-matic multilib` eclasses.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+SRC_URI="http://gdlp01.c-wss.com/gds/3/0100004693/01/${MY_PN}-source-${PV}-1.tar.gz"
+
+LICENSE="GPL-2"
+SLOT="${PV}"
```
This package doesn't look like you can install multiple versions simultaneously, so slotting is wrong.

mfld-fr replied:
Actually, yes, we should be able to install multiple versions of `cnijfilter-drivers`, as each printer model is supported only by a specific version. Only `cnijfilter` is common to all models and evolves through versions. See the discussion in the referenced bug. Ignored & for a next PR.

mgorny replied:
If that is indeed the case, you should really add an appropriate comment. However, I should point out that such a layout will be terribly hard to use for the users.

mfld-fr replied:
You are 100% right : this policy of Canon is really a nightmare for the users, and generate tons of discussions to find the right version for a given printer model, and to manage the obsolescence of that version. This is probably the reason why many guys try to merge all the versions in the latest (just look for projects with `cnijfilter` keyword on Github).

mfld-fr replied:
Closed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,112 @@
+# Copyright 1999-2012 Gentoo Foundation
```
Same comments as for the other package, for a start.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+# Distributed under the terms of the GNU General Public License v2
+# $Header: $
+
+EAPI=5
```
Go for EAPI 6.

mfld-fr replied:
Fixed

mfld-fr commented on net-print/cnijfilter-drivers/Manifest:
```diff
@@ -0,0 +1 @@
+DIST cnijfilter-source-3.80-1.tar.gz 8923054 SHA256 8b6d408f18191f19465ee8fc31aa08455e8bec186fdd3f02ee822f53a9b086a9 SHA512 95a16e3b4fc38ce0b7a12bd74466d97e726bc410b59bf6d1963fa52b16a8cc67f6a993a5ef945107201f860d8ac6734c462bc0bf6d2160d6c85c5f61aff040c1 WHIRLPOOL 2e27afa454ce1fa41700f65ace7ae7469464cc8685499c4927c559aef2fd79613297d1ec1e5cf1dd309da541c5b662a0c4d4d71fabf4b514e186eb2bceb80c5c
```
The package `cnijfilter-drivers` depends on `cnijfilter`, so moving `cnijfilter-drivers` in another branch for a next PR.

mgorny replied:
Not two pull requests. Two commits on a single branch.

mfld-fr replied:
OK, but let us complete the review first on `cnijfilter` for me to understand well your requirements, then I will apply them to `cnijfilter-drivers` and move it back on the line.

mfld-fr replied:
Fixed

mfld-fr commented:
Thanks for the review ! Now working on it...

mfld-fr commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,167 @@
+# Copyright 1999-2012 Gentoo Foundation
```
Fixed

mfld-fr commented:
@mgorny : fixed what I understood from your review in the first package `cnijfilter`. Could you tell me if more fixing is required on that first one, or if I could squash into a first commit and begin to work on the second package `cnijfilter-drivers` ?

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+
+inherit autotools
+
+PS="${PN}-source-${PV}-1"
```
We usually name this `MY_P`, and I suggest you follow suit for clarity.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+
+# Patches applied by the default src_prepare() since EAPI=6
+
+PATCHES=( \
```
Those backslashes are unnecessary here.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+}
+
+src_configure() {
+	_dir_build "${DIRS}" "eautoreconf"
```
Inline that loop from `_dir_build`, and don't print that `>>>` thing pretending to come from Portage. Also, `eautoreconf` belogs in `src_prepare`.

Alternatively, you could create a simple `configure.ac`+`Makefile.am` using [AC_CONFIG_SUBDIRS + SUBDIRS](https://www.gnu.org/software/automake/manual/html_node/Subpackages.html) that would let you build it like a regular package with autotools automatically recurring into subdirectories.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+}
+
+src_compile() {
+	_dir_build "${DIRS}" "emake"
```
Don't do `pushd`+`emake`, use `emake -C directory`.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+	_dir_build "${DIRS}" "emake DESTDIR=${D} install"
+
+	if use net; then
+		pushd com/${_libdir_pkg} > /dev/null
```
You're using this exactly once, so don't create an extra variable for it.

External commands need `||die`

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+
+src_install() {
+	local _libdir="${EPREFIX}/usr/$(get_libdir)"
+	local _libdir_pkg=libs_bin$(use amd64 && echo 64 || echo 32)
```
`usex amd64 64 32`

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+
+	if use net; then
+		pushd com/${_libdir_pkg} > /dev/null
+		dodir ${_libdir}
```
`dodir` needs a path *without* EPREFIX. But it's not necessary since…

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+		pushd com/${_libdir_pkg} > /dev/null
+		dodir ${_libdir}
+		# no doexe to preserve symlinks
+		cp -a libcnnet.so* "${D}/${_libdir}" || die
```
…`dolib` is the function you're looking for.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+}
+
+pkg_postinst() {
+	einfo ""
```
`einfo` is for status/progress messages that are of no value once the build is complete. You're lookin for `elog`.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+
+pkg_postinst() {
+	einfo ""
+	einfo "For installing a printer:"
```
That sounds like entirely boilerplate 'how to use CUPS' instruction. Any reason you're repeating it here?

mfld-fr replied:
Removed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,104 @@
+	einfo "   -> Printers -> Add Printer"
+	einfo ""
+	einfo "If you experience any problem, please visit:"
+	einfo " http://forums.gentoo.org/viewtopic-p-3217721.html"
```
https. Besides, Forums threads are not really means of providing support and/or information. They have very poor SNR.

mfld-fr replied:
Removed

mfld-fr commented:
@mgorny : Thanks for that second review ! Now working on it...

mfld-fr commented:
@mgorny : I fixed all the points from your second review. Now waiting for your green light to proceed further...

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+	"${FILESDIR}/cnijfilter-3.80-cups1.6.patch"
+	)
+
+pkg_setup() {
```
As a minor note, you could move this logic to `src_prepare` (or surround with `${MERGE_TYPE}` conditional) since I don't think it's needed when you install binary packages.

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+	local d
+	for d in ${DIRS}; do
+		pushd ${d} >/dev/null || die
+		eautoreconf ${d}
```
`eautoreconf` doesn't take any parameters.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+}
+
+src_configure() {
+
```
ELEADINGEMPTYLINE

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+
+src_configure() {
+
+	# progpath must be set otherwise defaulted to /usr/local/bin
```
I think it would fit better next to `econf`

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+
+	local d
+	for d in ${DIRS}; do
+		pushd ${d} >/dev/null || die
```
Always quote paths. Even if by design they can't contain spaces. In fact, you should convert whole `DIRS` to bash array, and use `"${DIRS[@]}"` to respect quoting.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+	local d
+	for d in ${DIRS}; do
+		pushd ${d} >/dev/null || die
+		econf --enable-progpath=${EPREFIX}/usr/bin
```
EPREFIX may contain spaces, quote it.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+src_install() {
+	local d
+	for d in ${DIRS}; do
+		emake -C ${d} DESTDIR=${D} install
```
Also missing quoting around `${D}`.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,96 @@
+	if use net; then
+		pushd com/libs_bin$(usex amd64 64 32) >/dev/null || die
+		dolib.so libcnnet.so.1.2.2
+		dolib.so libcnnet.so
```
Any reason not to use `libcnnet.so*`?

mfld-fr replied:
Fixed

mfld-fr commented:
@mgorny : Thanks for that third review ! Now working on it...

mfld-fr commented:
@mgorny : I fixed all the points from your third. Now waiting for your feedback...

mgorny commented on net-print/cnijfilter/cnijfilter-3.80.ebuild:
```diff
@@ -0,0 +1,95 @@
+		DIRS=(libs pstocanonij backend)
+		use net && DIRS+=(backendnet)
+		use servicetools && DIRS+=(cngpij)
+		fi
```
Misindent.

mfld-fr replied:
Fixed

mfld-fr commented:
About your side note: actually the two packages build from the same set of sources, but the first builds a single variant, and the second builds as many variants as printer models. This is why in the bug discussion, a Gentoo developer (scarabeus) splitted that package into two parts. But as the sources are the same for both, we still need to apply the same patches.

mfld-fr commented:
Squashed all the previous commits into a single one for the first package `cnijfilter` (common part). Now reworking the second package `cnijfilter-drivers` (printer-specific part)...

mfld-fr commented:
@mgorny : I moved back the second package `cnijfilter-drivers` and cleaned up according to your previous reviews. Now I need your help to factorize the current code with the multibuild.eclass, because I am a beginner in writing ebuilds. Could you please provide me some guidelines to do so ?

mgorny commented:
You use it like this:

1. Set `MULTIBUILD_VARIANTS` to an array of variants you'd like to build, i.e. different printer models.
2. Call `multibuild_foreach_variant somefunc` to call `somefunc` for each variant.

where `somefunc` would appropriately be your functions doing different phases of the ebuild. The eclass will set `BUILD_DIR` which you can use to do the building, i.e. given your previous code for configure you'd do something like:

```
my_configure() {
  for x in subdir1 subdir2 subdir3...; do
    mkdir -p "${BUILD_DIR}/${x}" || die
    cd "${BUILD_DIR}/${x}" || die
    ECONF_SOURCE=${S}/${x} econf ...
  done
}
```

This will trigger an out-of-source configure, i.e. it will use sources from another directory to avoid unnecessary copying. For other phases, you'd do something like:

```
my_compile() {
  for x in subdir1 subdir2 subdir3...; do
    emake -C "${BUILD_DIR}/${x}"
  done
}
```

and so on.

mfld-fr commented:
@mgorny : Thanks for your guidelines to use `multibuild.eclass`. I switched to that, could you please review again and tell me if my usage is correct ?

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,148 @@
+	[[ -z ${LINGUAS} ]] && LINGUAS="en"
+
+	if [[ ${MERGE_TYPE} == source ]]; then
+
```
Avoid empty lines on top of block.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,148 @@
+
+		# select variants
+
+		for (( i=0; i<${#PRINTERS[@]}; i++ )); do
```
The whole point was that you don't have to do the `(( ... ))` anymore and you can just do `for x in "${PRINTERS[@]}"`

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,148 @@
+
+		for (( i=0; i<${#PRINTERS[@]}; i++ )); do
+			local name=${PRINTERS[$i]#*:}
+			use ${name} && MULTIBUILD_VARIANTS+=(${name})
```
Quoting.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,148 @@
+	# missing macros directory make aclocal fail
+	mkdir maintenance/m4 || die
+
+	# maybe default could be called here ?
```
Yes, it could.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,148 @@
+	for d in "${DIRS[@]}"; do
+		mkdir -p "${BUILD_DIR}/${d}" || die
+		cd "${BUILD_DIR}/${d}" || die
+    	ECONF_SOURCE=${S}/${d} econf \
```
Misindent.

mfld-fr replied:
Fixed

mfld-fr commented:
@mgorny : I fixed all the points from your latest review. Any more fix needed or could I squash all the commits related to the `cnijfilter-drivers` package ?

mfld-fr commented:
I have just had a thought: should the `printing@gentoo.org` project be added as intermediary maintainer in the `metadata.xml` files ?

mfld-fr commented:
Adding @gentoo/printing in the loop as parent project (?) for this package.

mfld-fr commented:
@gento/proxy-maint : gentle ping

mfld-fr commented:
@mgorny : gentle ping

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+		local p
+		for p in "${PRINTERS[@]}"; do
+			local name="${p#*:}"
+			use ${name} && MULTIBUILD_VARIANTS+=("${name}")
```
Minor: missing quoting around first `${name}`. It'd be nice to quote consistently.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+	local _libdir_pkg=libs_bin$(usex amd64 64 32)
+	local _ppddir="${EPREFIX}/usr/share/cups/model"
+
+	dodir ${_libdir}
```
`dodir` prepends `${EPREFIX}` itself, so you need to pass the unprefixed path here.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+
+	dodir ${_libdir}
+	# create symlink from cnijlib to bjlib as some formats need it
+	dosym ${_libdir}/cnijlib ${_libdir}/bjlib
```
So does `dosym` for its second argument. For the first argument, please use relative path instead. That'd be just `cnijlib` if I'm not mistaken.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+	dosym ${_libdir}/cnijlib ${_libdir}/bjlib
+
+	local i
+	for (( i=0; i<${#PRINTERS[@]}; i++ )); do
```
Switch this one to `for ... in` as well, please.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+	for (( i=0; i<${#PRINTERS[@]}; i++ )); do
+		local name="${PRINTERS[$i]#*:}"
+		local pid="${PRINTERS[$i]%:*}"
+		if use ${name}; then
```
Minor: quoting.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+		local pid="${PRINTERS[$i]%:*}"
+		if use ${name}; then
+			dolib.so ${pid}/${_libdir_pkg}/*
+			insinto "${_libdir}/cnijlib"
```
Spurious EPREFIX again.

mfld-fr replied:
Fixed

mgorny commented on net-print/cnijfilter-drivers/cnijfilter-drivers-3.80.ebuild:
```diff
@@ -0,0 +1,144 @@
+			dolib.so ${pid}/${_libdir_pkg}/*
+			insinto "${_libdir}/cnijlib"
+			doins ${pid}/database/*
+			insinto "${_ppddir}"
```
Likewise.

mfld-fr replied:
Fixed

mgorny commented:
I've thought about it some, and to be honest I don't think the slotting solution will fly. It will cause a lot of confusion to the users, and some of them will suddenly lose a driver on upgrade+depclean when you add a new version. Most of the people would not even think about looking at old versions.

i will understand if that's too much work for you but I think this would be best to do as one-package-per-printer.

mfld-fr commented:
@mgorny : so you want the same layout as in AUR (https://repology.org/metapackages/all/?search=cnijfilter) ?

If factorization is removed, there is roughly one Canon release per year (2012-2017), each release supporting about 6~8 printer series, so one would have to manage about ~35 packages, not including the backward with about 5 years more (to support at least the ~10 years printer common mechanical lifetime), total ~70 packages. Plus the already existing workload to maintain ~10 set of patches.

This could only be managed through automation, like generating the `ebuild` from a common template, and a table of the printer series, manufacturer's source packages and patches. Otherwise it would become the mess, with many maintainers (up to one for each printer series) and no consistency, as in AUR today.

I agree it would be easier for the dummy user, but is Gentoo targeting such user profile ? To help selecting the right version for each printer serie, I also started to write a Wiki page : https://wiki.gentoo.org/index.php?title=Canon_Pixma_Printer.

This is an important design decision, so please confirm before I proceed. I can commit to maintain a `cnijfilter-common` and a `cnijfilter-ip7200`, because I can fully test it with my IP 7250 printer and be confident about the package quality. But for other models, one would have to rely on other contributors in testing the packages, and I could only propose package candidates.

mfld-fr commented:
Also asking opinion from @tokiclover and @sabayon, as experienced contributors on this topic.

mgorny commented:
Yes, that's the correct layout. And as you said, those will probably be low maintenance packages, with a lot of common code that you could easily automate locally. As for the patches, I'd suggest using a common patch tarball. If that's the final patch set, one of us can make a tarball of them and put it on Gentoo hosting for you.

Gentoo is targeting good quality of design and using the newest available version of software whenever possible. A design which relies on user having to do research and force old version of software to make it work is simply wrong.

mfld-fr commented:
@mgorny : I removed the multibuild from the specific-part and kept only the printer series which I am able to test (ip7200). I also squashed the previous commits while renaming `cnijfilter-drivers` to `cnijfilter-ip7200`. Could you please review this new layout ?

mgorny commented on net-print/cnijfilter-common/cnijfilter-common-3.80.ebuild:
```diff
@@ -0,0 +1,95 @@
+	"${FILESDIR}/cnijfilter-3.70-ppd2.patch"
+	"${FILESDIR}/cnijfilter-3.70-libexec-cups.patch"
+	"${FILESDIR}/cnijfilter-3.70-libexec-backend.patch"
+	"${FILESDIR}/cnijfilter-3.80-cups1.6.patch"
```
A minor request but could you try to use a more meaningful patch names? Something that could let others figure out which patches they need for their drivers because right now, I don't understand what's the difference e.g. between 'cups1.6' and 'cups' patches.

mfld-fr replied:
Already explained in a previous review log the practical reason why I am keeping the patches as they are sourced from many sources (> 10 related bugs, 2 layouts and 5 other GH repos). I am currently consolidating them in my own working copy of the sources, and I will submit more consistent (+ automation compatible) patches in a next release. Closed.

mfld-fr commented on net-print/cnijfilter-common/cnijfilter-common-3.80.ebuild:
```diff
@@ -0,0 +1,95 @@
+pkg_setup() {
+	[[ -z ${LINGUAS} ]] && LINGUAS="en"
+
+	if [[ ${MERGE_TYPE} == source ]]; then
```
Fixed

mfld-fr commented on net-print/cnijfilter-ip7200/cnijfilter-ip7200-3.80.ebuild:
```diff
@@ -0,0 +1,113 @@
+
+	dodir ${_libdir}
+	# create symlink from cnijlib to bjlib as some formats need it
+	dosym cnijlib ${_libdir}/bjlib
```
Fixed

mgorny commented on net-print/cnijfilter-ip7200/cnijfilter-ip7200-3.80.ebuild:
```diff
@@ -0,0 +1,113 @@
+SRC_URI="http://gdlp01.c-wss.com/gds/3/0100004693/01/${MY_P}.tar.gz"
+
+LICENSE="GPL-2"
+
```
Nit: stray empty line.

mfld-fr replied:
Fixed

mfld-fr commented:
Sorry, but this time it looks not reasonable to follow you. The submitted `ebuild` are working well, we spent already 3 weeks to iterate, and now you request lately a design change by packing the patches, that would be temporary in the end, as I said that I will deliver soon a next revision (-r1) with a consolidated patch that will do the same as your packing.

I am also working currently on the removal of the current compilation warnings that have actually no impact (no problem on the real printer since 2 years now), and I am kept stuck on this PR in place of going forward on that effective quality improvement.

I therefore won't change any more these package candidates for that first revision. Please accept as is, reject it or implement your packing yourselves, because I don't think packing the patches will change the game.

Thanks anyway for your time, I learned much in writting `ebuild`.

urcindalo commented:
I really appreciate your efforts, @mfld-fr 
Thanks to you I now have a working Canon PIXMA iP7250 on my Gentoo box.  I hope your ebuild enters the official tree soon.

mfld-fr commented:
@urcindalo : nice comment, thank you !

mfld-fr commented:
@mgorny : Yes, I already spotted them. The ones related to --enable-progpath are fixed inside `ebuild` in the next package release. But for others (configure.in, CFLAGS and LDFLAGS), fixing requires to rework many Canon scripts & sources, not just patching a few files.

Canon did only one release of that 3.80 sources, and provide no way to contribute to the product by submitting changes. Therefore, one will never be able to fix the (configure.in, CFLAGS, LDFLAGS) in the mainstream, especially for /usr/lib64/libcnnet.so.1.2.2, as that library and some others are provided as binary files in their source package.

So either you accept these limitations and allow the Gentoo users to work with their Canon printers without going through another box (Windows ?..), because these packages are proven to work very well on site despite these QA issues; or you reject the whole.

IMHO, I would more prioritize my users needs and open the gate for continuous improvement, than trying to reach the perfection. The "voice of your customers" looks very clear by just searching for the bugs with the "cnijfilter" keyword  and then reading. Up to you to listen that voice or to ignore it, but in the latest, don't cry if other distros take the market lead by making the good balancing...

mgorny commented:
I would really appreciate if you stopped the preaching and focused on doing your work. As for prebuilt files, there is a `QA_PREBUILT` variable (`man 5 ebuild`) for that purpose.

mfld-fr commented:
My work ?!? I am not paid by Gentoo to become an expert like you in the portage system. Thank you for the trick about prebuilt binaries that I will use in my next release, but I am afraid it won't change the whole picture.

a17r commented:
@mfld-fr: Sure, we all work together on this platform and are not paid for that. Please also appreciate the time that it takes another developer to review your submissions.

While a package hardly ever enters tree to be called perfect, when we know there are issues, and we know the way to fix them, then we do that.

mfld-fr commented:
@a17r : Be sure I think that one could never thanks enough all the people that work hard to maintain our favorite distro and to help newbies like me to improve their contribution.

But the keypoint here is that the mentionned QA issues have stricly no impact on the final user, and thus cannot be a blocker to deliver a working feature. First times of that review had added value in the sense it improved the ebuild itslef and the use case, but now what is asked is more to work on the Canon software itself, or to hide its defects, beyond the scope of a smooth integration. I am not the maintainer of the Canon software, just a user that need it working in Gentoo, as the others listed in the related bugs.

Such QA issues can be fixed with a lower priority (this is what I am doing in the background). As in software industry, one should keep in mind that time to market with a good level of quality is critical for customer satisfaction, not looking for a perfect product that would come too late. Canon drivers have been requested for more than 10 years now, don't you think it is time to stop playing this game ?

mgorny commented:
So are you saying that a warning telling *user* that in the future *the autoconf will explode* has no impact on the final user? Having user be shot with a screenful of warnings and spending his effort reporting a bug that shouldn't be there in the first place is not an impact? Or when things actually stop working, will that have no impact on the user?

My problem is that you submit an ebuild with known issues, and instead of either fixing them or asking about them, you pretend that they are not there. Then I verify it and suddenly discover that you have been ignoring a screenful of issues all the time.

mgorny commented on net-print/cnijfilter-ip7200/cnijfilter-ip7200-3.80.ebuild:
```diff
@@ -0,0 +1,113 @@
+HOMEPAGE="https://wiki.gentoo.org/wiki/Canon_Pixma_Printer"
+SRC_URI="http://gdlp01.c-wss.com/gds/3/0100004693/01/${MY_P}.tar.gz"
+
+LICENSE="GPL-2"
```
Are you convinced this is GPL-2? Because AFAICS there's an EULA in `LICENSE*` which seems to cover the prebuilt files.

mfld-fr replied:
No, you are right, there is a Canon prologue before the GPL v2 not only in the license files, but also displayed on the download web pages. Should i submit that prologue as a new license in Gentoo ?

mfld-fr commented:
@mgorny : Have you read my comments in the referenced bug and in this PR header ? I always stated that there are build warnings and QA issues. Sorry if you discover it only now. And do you think I never inspected these warnings to be sure they won't be a potential problem that would impact me the first, as the printer daily user ?

About the impact, I repeat again : these can be fixed in a second step, before the big crunch you are predicting in the future. I had a look to many bugs reports for some packages I am interested in : users are usually more concerned by package availability and working state than a clean build or emerge. Ones that open bugs for QA are... Larry and developers ! Think also about that.

QA should be a helper for us to continuously improve the quality, not a blocker. And yes, you are right, I am preaching to not see again the damage of a misused QMS on too many businesses.

a17r commented:
No, users will be concerned by flashing red QA warnings, and they do submit bugs. Thing is, this could have been fixed instead of spending time on several paragraphs. There's just no reason not to do it, even if you only get around to do it tomorrow or the day after - that won't matter after waiting 10 years, as you said.

mfld-fr commented:
@a17r : thing is, we all together spent too much time for almost one month just on one package, and I am now afraid by the so predictible amount of time it will require to push more.

Plus the fact that there is obviously a disconnect in our idea of quality management in non-critical software area. Thanks anyway for your time and this conversation, it helped me to realize it won't match as a contributor, and you saved much time of everybody in the end :-).

Please close this PR as rejected, I keep the submission branch in case anyone would like to use the package after it was improved during the review.
